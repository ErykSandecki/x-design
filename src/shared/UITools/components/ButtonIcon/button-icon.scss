.ButtonIcon {
  @include theme-selector(
    'neutral-3',
    background-color,
    '&:hover:not(.ButtonIcon--selected)'
  );

  align-items: center;
  border-radius: 5px;
  cursor: pointer;
  display: flex;
  height: 24px;
  justify-content: center;
  width: 24px;

  &--selected {
    &--light {
      background-color: rgba($color: $--xd-blue-1--light, $alpha: 30%);
    }

    &--dark {
      background-color: rgba($color: $--xd-blue-1--dark, $alpha: 30%);
    }
  }
}
