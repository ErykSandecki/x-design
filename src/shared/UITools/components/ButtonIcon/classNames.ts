// utils
import { composeClassNames } from 'utils';

export const className = 'ButtonIcon';

export const classNames = composeClassNames(className, [className, 'selected']);
