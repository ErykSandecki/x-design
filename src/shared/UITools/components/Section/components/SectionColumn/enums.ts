export enum GridColumnType {
  single = 'single',
  twoInputs = 'twoInputs',
}
