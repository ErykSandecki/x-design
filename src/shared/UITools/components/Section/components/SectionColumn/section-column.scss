$--xd-gap: 10px;

.SectionColumn {
  &--with-margin {
    margin-bottom: 7.5px;
  }

  &__labels {
    column-gap: $--xd-gap;
  }

  &__label {
    @include theme('neutral-2', color);

    font-size: 9px;
    margin-bottom: 5px;
  }

  &__content {
    column-gap: $--xd-gap;

    &--single {
      grid-template-columns: 1fr;
    }

    &--two-inputs {
      grid-template-columns: repeat(2, 1fr);
    }
  }
}
