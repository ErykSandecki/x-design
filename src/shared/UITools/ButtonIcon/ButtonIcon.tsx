import { FC, useRef } from 'react';
import { noop } from 'lodash';

// components
import Box, { TBoxProps } from '../../UI/Box/Box';
import ButtonIconPopover from './ButtonIconPopover/ButtonIconPopover';
import Icon, { TIconProps } from '../../UI/Icon/Icon';
import Tooltip, { TTooltipProps } from '../../UI/Tooltip/Tooltip';

// hooks
import { useClickEvent } from './hooks/useClickEvent';
import { useOutsideClick, useTheme } from 'hooks';

// others
import { className, classNames } from './classNames';

// styles
import styles from './button-icon.scss';

// types
import { ColorsTheme, E2EAttribute } from 'types';
import { TE2EDataAttributeProps } from '../../E2EDataAttributes/E2EDataAttribute';
import { TPopoverProps } from '../Popover/Popover';

export type TSectionProps = Pick<TIconProps, 'name'> &
  TBoxProps & {
    e2eValue?: TE2EDataAttributeProps['value'];
    idContainer?: string;
    popoverChildren?: TPopoverProps['children'];
    popoverOffset?: TPopoverProps['offset'];
    selected?: boolean;
    tooltip?: Omit<TTooltipProps, 'children'>;
  };

export const ButtonIcon: FC<TSectionProps> = ({
  e2eValue = '',
  idContainer,
  name,
  onClick,
  popoverChildren,
  popoverOffset,
  selected = false,
  tooltip,
  ...restProps
}) => {
  const ref = useRef(null);
  const { classNamesWithTheme, cx } = useTheme(classNames, styles);
  const { selected: selectedPopover, setSelected: setSelectedPopover } = useOutsideClick([ref], ref, noop, idContainer);
  const isSelected = selectedPopover || selected;
  const onClickHandler = useClickEvent(onClick, selectedPopover, setSelectedPopover);

  return (
    <Tooltip ref={ref} {...tooltip}>
      <Box
        classes={{
          className: cx(classNamesWithTheme[className].name, [
            classNamesWithTheme[className].modificators.selected,
            isSelected,
          ]),
        }}
        e2eAttribute={E2EAttribute.buttonIcon}
        e2eValue={e2eValue}
        onClick={onClickHandler}
        {...restProps}
      >
        <Icon color={selected ? ColorsTheme.blue1 : ColorsTheme.neutral1} height={14} name={name} width={14} />
        {popoverChildren && (
          <ButtonIconPopover
            ref={ref}
            selected={selectedPopover}
            setSelected={setSelectedPopover}
            offset={popoverOffset}
          >
            {popoverChildren}
          </ButtonIconPopover>
        )}
      </Box>
    </Tooltip>
  );
};

export default ButtonIcon;
