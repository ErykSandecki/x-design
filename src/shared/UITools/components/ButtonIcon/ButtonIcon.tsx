import { FC } from 'react';

// components
import Box, { TBoxProps } from '../../../UI/components/Box/Box';
import Icon, { TIconProps } from '../../../UI/components/Icon/Icon';
import Tooltip, { TTooltipProps } from '../../../UI/components/Tooltip/Tooltip';

// hooks
import { useTheme } from 'hooks';

// others
import { className, classNames } from './classNames';

// styles
import styles from './button-icon.scss';

// types
import { ColorsTheme, E2EAttribute } from 'types';
import { TE2EDataAttributeProps } from '../../../E2EDataAttributes/E2EDataAttribute';

export type TSectionProps = Pick<TIconProps, 'name'> &
  TBoxProps & {
    e2eValue?: TE2EDataAttributeProps['value'];
    selected?: boolean;
    tooltip?: Omit<TTooltipProps, 'children'>;
  };

export const ButtonIcon: FC<TSectionProps> = ({ e2eValue = '', name, selected = false, tooltip, ...restProps }) => {
  const { classNamesWithTheme, cx } = useTheme(classNames, styles);

  return (
    <Tooltip {...tooltip}>
      <Box
        classes={{
          className: cx(classNamesWithTheme[className].name, [
            classNamesWithTheme[className].modificators.selected,
            selected,
          ]),
        }}
        e2eAttribute={E2EAttribute.buttonIcon}
        e2eValue={e2eValue}
        {...restProps}
      >
        <Icon color={selected ? ColorsTheme.blue1 : ColorsTheme.neutral1} height={14} name={name} width={14} />
      </Box>
    </Tooltip>
  );
};

export default ButtonIcon;
