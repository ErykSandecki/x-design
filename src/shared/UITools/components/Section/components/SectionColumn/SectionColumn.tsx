import { FC, ReactNode } from 'react';
import { size } from 'lodash';

// components
import Box from '../../../../../UI/components/Box/Box';
import { Small } from '../../../../../UI/components/Typography';

// hooks
import { useTheme } from 'hooks';

// others
import { className, classNames } from './classNames';

// styles
import styles from './section-column.scss';

// types
import { E2EAttribute } from 'types';
import { GridColumnType } from './enums';

export type TSectionColumnProps = {
  buttonsIcon?: Array<ReactNode>;
  children: ReactNode;
  gridColumnType?: GridColumnType;
  labels?: [string] | [string, string];
  withMargin?: boolean;
};

export const SectionColumn: FC<TSectionColumnProps> = ({
  buttonsIcon = [],
  children,
  gridColumnType = GridColumnType.single,
  labels = [],
  withMargin = false,
}) => {
  const { classNamesWithTheme, cx } = useTheme(classNames, styles);
  const buttonsWidthTotal = (size(buttonsIcon) || 1) * 24;
  const additionalGap = size(buttonsIcon) === 2 ? 2.5 : 0;
  const width = `calc(100% - ${buttonsWidthTotal}px - ${additionalGap}px - 8px)`;

  return (
    <Box
      classes={{
        className: cx(classNamesWithTheme[className].name, [
          classNamesWithTheme[className].modificators.withMargin,
          withMargin,
        ]),
      }}
      e2eAttribute={E2EAttribute.section}
    >
      <Box classes={{ className: cx(classNamesWithTheme.labels) }} style={{ width }} sx={{ display: 'flex' }}>
        {!!size(labels) &&
          labels.map((label, index) => (
            <Small classes={{ className: cx(classNamesWithTheme.label) }} key={index}>
              {label}
            </Small>
          ))}
      </Box>
      <Box sx={{ alignItems: 'center', columnGap: '8px', display: 'flex' }}>
        <Box
          classes={{
            className: cx(classNamesWithTheme.content.name, classNamesWithTheme.content.modificators[gridColumnType]),
          }}
          style={{ width }}
          sx={{ display: 'grid' }}
        >
          {children}
        </Box>
        <Box sx={{ alignItems: 'center', columnGap: '2.5px', display: 'flex', flexGrow: 1, justifyContent: 'center' }}>
          {buttonsIcon.map((buttonIcon) => buttonIcon)}
        </Box>
      </Box>
    </Box>
  );
};

export default SectionColumn;
