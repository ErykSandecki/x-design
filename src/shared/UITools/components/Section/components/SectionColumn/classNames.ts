import { kebabCase, mapValues } from 'lodash';

// types
import { GridColumnType } from './enums';

export const className = 'SectionColumn';

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      withMargin: `${className}--with-margin`,
    },
  },
  labels: `${className}__labels`,
  label: `${className}__label`,
  wrapper: `${className}__wrapper`,
  content: {
    name: `${className}__content`,
    modificators: {
      ...mapValues(GridColumnType, (gridColumnType) => `${className}__content--${kebabCase(gridColumnType)}`),
    },
  },
  buttons: `${className}__buttons`,
};
