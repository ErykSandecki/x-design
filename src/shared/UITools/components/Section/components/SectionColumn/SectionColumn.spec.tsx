import { render } from '@testing-library/react';

// components
import ButtonIcon from 'shared/UITools/components/ButtonIcon/ButtonIcon';
import SectionColumn from './SectionColumn';

describe('SectionColumn snapshots', () => {
  it('should render SectionColumn', () => {
    // before
    const { asFragment } = render(<SectionColumn>children</SectionColumn>);

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with buttons', () => {
    // before
    const { asFragment } = render(
      <SectionColumn buttonsIcon={[<ButtonIcon name="Close" key={0} />]}>children</SectionColumn>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with buttons and input connector', () => {
    // before
    const { asFragment } = render(
      <SectionColumn buttonsIcon={[<ButtonIcon name="Close" key={0} />]} inputConnector>
        children
      </SectionColumn>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with buttons & labels', () => {
    // before
    const { asFragment } = render(
      <SectionColumn buttonsIcon={[<ButtonIcon name="Close" key={0} />]} labels={['label']}>
        children
      </SectionColumn>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
