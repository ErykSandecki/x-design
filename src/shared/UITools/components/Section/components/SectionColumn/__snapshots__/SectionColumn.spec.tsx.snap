// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionColumn snapshots should render SectionColumn 1`] = `
<DocumentFragment>
  <div
    class="css-1jnxvbp SectionColumn SectionColumn--light"
    data-test-section=""
  >
    <div
      class="css-xdzdnq SectionColumn__labels SectionColumn__labels--light"
      data-test-box=""
      style="width: calc(100% - 24px - 0px - 8px);"
    />
    <div
      class="css-1r00qlg"
      data-test-box=""
    >
      <div
        class="css-1m2m6pt SectionColumn__content SectionColumn__content--light SectionColumn__content--single SectionColumn__content--single--light"
        data-test-box=""
        style="width: calc(100% - 24px - 0px - 8px);"
      >
        children
      </div>
      <div
        class="css-c1qms7"
        data-test-box=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`SectionColumn snapshots should render with buttons & labels 1`] = `
<DocumentFragment>
  <div
    class="css-1jnxvbp SectionColumn SectionColumn--light"
    data-test-section=""
  >
    <div
      class="css-xdzdnq SectionColumn__labels SectionColumn__labels--light"
      data-test-box=""
      style="width: calc(100% - 24px - 0px - 8px);"
    >
      <small
        class="css-1jnxvbp SectionColumn__label SectionColumn__label--light Typography Typography--light Typography--neutral1 Typography--neutral1--light Typography--regular Typography--regular--light Typography--normal Typography--normal--light Typography--small Typography--small--light"
        data-test-text=""
        style="text-align: inherit;"
      >
        label
      </small>
    </div>
    <div
      class="css-1r00qlg"
      data-test-box=""
    >
      <div
        class="css-1m2m6pt SectionColumn__content SectionColumn__content--light SectionColumn__content--single SectionColumn__content--single--light"
        data-test-box=""
        style="width: calc(100% - 24px - 0px - 8px);"
      >
        children
      </div>
      <div
        class="css-c1qms7"
        data-test-box=""
      >
        <div
          class="css-1jnxvbp ButtonIcon ButtonIcon--light"
          data-test-button-icon=""
        >
          <div
            class="css-1jnxvbp Icon Icon--light Icon--neutral1 Icon--neutral1--light"
            data-test-icon="close"
            height="14"
            name="Close"
            width="14"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SectionColumn snapshots should render with buttons 1`] = `
<DocumentFragment>
  <div
    class="css-1jnxvbp SectionColumn SectionColumn--light"
    data-test-section=""
  >
    <div
      class="css-xdzdnq SectionColumn__labels SectionColumn__labels--light"
      data-test-box=""
      style="width: calc(100% - 24px - 0px - 8px);"
    />
    <div
      class="css-1r00qlg"
      data-test-box=""
    >
      <div
        class="css-1m2m6pt SectionColumn__content SectionColumn__content--light SectionColumn__content--single SectionColumn__content--single--light"
        data-test-box=""
        style="width: calc(100% - 24px - 0px - 8px);"
      >
        children
      </div>
      <div
        class="css-c1qms7"
        data-test-box=""
      >
        <div
          class="css-1jnxvbp ButtonIcon ButtonIcon--light"
          data-test-button-icon=""
        >
          <div
            class="css-1jnxvbp Icon Icon--light Icon--neutral1 Icon--neutral1--light"
            data-test-icon="close"
            height="14"
            name="Close"
            width="14"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
